keps/sig-network/0030-nodelocal-dns-cache.md

```diff
@@ -86,45 +86,109 @@ The node's resolv.conf will be used by this local agent for all other cache miss
 
 #### Daemonset, Service, Listen Interface for caching agent
 
-The caching agent daemonset runs in hostNetwork mode in kube-system namespace with a Priority Class of “system-node-critical”. It listens for dns requests on a dummy interface created on the host. A separate ip address is assigned to this dummy interface, so that requests to kube-dns or any other custom service are not incorrectly intercepted by the caching agent. This ip address is obtained by creating a nodelocaldns service, with no endpoints. Kubelet takes the service name as an argument `--localDNS=<namespace>/<svc name>`, looks up the ip address and populates pods' resolv.conf with this value instead of clusterDNS. Each cluster node will have a dummy interface with this service ip assigned to it. This IP will be handled specially because of the NOTRACK rules described in the section below.
+The caching agent daemonset runs in hostNetwork mode in kube-system namespace with a Priority Class of “system-node-critical”. It listens for dns requests on a dummy interface created on the host. A separate ip address is assigned to this dummy interface, so that requests to kube-dns or any other custom service are not incorrectly intercepted by the caching agent. This ip address is obtained by creating a nodelocaldns service, with same endpoints as clusterDNS ip. Using the same endpoints as clusterDNS helps reduce DNS downtime in case of upgrades/restart. When no other special handling is provided, queries to the nodelocaldns ip will be served by kube-dns/coreDNS pods. Kubelet takes the service name as an argument `--localDNS=<namespace>/<svc name>`, looks up the ip address and populates pods' resolv.conf with this value instead of clusterDNS. Each cluster node will have a dummy interface with this service ip assigned to it.
+
+Since this approach involves assigning a service IP to a dummy interface, it will not work in ipvs mode of kube-proxy. This is because kube-proxy creates a dummy interface bound to all service IPs in ipvs mode and ipvs rules are added to load-balance between endpoints. The packet seems to get dropped if there are no endpoints. If there are endpoints, adding iptables rules does not bypass the ipvs loadbalancing rules. Instead of using a service ip, a link-local ip can be selected by the user in case of ipvs mode. The --localDNS flag can be used to pass on either an ip address or a namespace/service name to support this. Kubelet will learn about the nodelocaldns ip from this flag in both cases.
```

Comment:
@m1093782566 Any other ideas on what we can do in ipvs mode?

Reply:
we could do that, but  we don't have a good fallback option, like you pointed out.

Reply:
Can't think of a less hacky way to get the service IP approach to work. This would be viable if the kube-proxy implementation supported a nodelocal type service. 
I will change this to listen on a user-specified link-local ip and reuse the --cluster-dns kubelet flag.